# xomx-tutorials
Tutorials for xomx ([https://github.com/perrin-isir/xomx](https://github.com/perrin-isir/xomx))

*List of colab notebooks:*
- *xomx_kidney_classif*: a tutorial in two phases:  
  [xomx_kidney_classif_1.ipynb](https://colab.research.google.com/github/perrin-isir/xomx-tutorials/blob/main/tutorials/xomx_kidney_classif_1.ipynb)
(phase 1) and  
  [xomx_kidney_classif_2.ipynb](https://colab.research.google.com/github/perrin-isir/xomx-tutorials/blob/main/tutorials/xomx_kidney_classif_2.ipynb) 
(phase 2).  
Remark: the phase 1, about importation and basic preprocessing of the data, can be skipped.  
*Goal of the tutorial:*  use a 
recursive feature elimination method on RNA-seq data to identify gene 
biomarkers for the differential diagnosis of three types of kidney cancer


- [xomx_pbmc.md](https://colab.research.google.com/github/perrin-isir/xomx-tutorials/blob/main/tutorials/xomx_pbmc.ipynb)  
*Goal of the tutorial:* follow the single cell RNA-seq [Scanpy tutorial on 3k PBMCs](
https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html), except
for the computation of biomarkers for which recursive feature elimination is used.

